enclave/src/polkadex_gateway.rs

```diff
@@ -0,0 +1,137 @@
+                    polkadex_balance_storage::reserve_balance(&main_account, order.market_id.base, order.quantity)?;
+                }
+            }
+        }
+        OrderType::FillOrKill | OrderType::PostOnly => {
+            error!("OrderType is not implemented");
+            return Err(sgx_status_t::SGX_ERROR_UNEXPECTED);
+        }
+    }
+    // Store the order
+    // TODO: we need OrderUUID for storing the order, it is given by Openfinex but we need OrderUUID to
+    // TODO: store the order in the orderbook storage
+    // polkadex_orderbook_storage::OrderbookStorage::add_order();
+    // TODO: Send order to Openfinex for inclusion
+    Ok(())
+}
+
+/// Place order function does the following
+/// 1. authenticate
+/// 2. send cancel_order to OpenFinex API
+/// 3. remove order from orderbook mirror
+/// 4. free reserved balance for the remainder of the order (in case of partial execution)
+/// 5. report result to sender
+pub fn cancel_order(main_account: AccountId, proxy_acc: Option<AccountId>, order_uuid: OrderUUID) -> SgxResult<()> {
+    // Authenticate
+    authenticate_user(main_account.clone()  , proxy_acc)?;
+    // TODO: Send cancel order to Openfinex API
+    // Mutate Balances
+    match polkadex_orderbook_storage::remove_order(&order_uuid)? {
+        Some(cancelled_order) => {
+            match cancelled_order.order_type {
+                OrderType::LIMIT => {
+                    if let Some(price) = cancelled_order.price {
+                        match cancelled_order.side {
+                            OrderSide::BID => {
+                                // TODO: Is it safe to use saturating_mul here?
+                                let amount = price.saturating_mul(cancelled_order.quantity);
+                                polkadex_balance_storage::unreserve_balance(main_account.clone(), cancelled_order.market_id.quote, amount)?;
+                            }
+                            OrderSide::ASK => {
+                                // TODO: Is it safe to use saturating_mul here?
+                                polkadex_balance_storage::unreserve_balance(main_account.clone(), cancelled_order.market_id.base, cancelled_order.quantity)?;
+                            }
+                        }
+                    } else {
+                        error!("Unable to find price for limit order");
+                        return Err(sgx_status_t::SGX_ERROR_UNEXPECTED);
+                    }
+                }
+                OrderType::MARKET => {}
+                OrderType::FillOrKill | OrderType::PostOnly => {
+                    error!("OrderType is not implemented");
+                    return Err(sgx_status_t::SGX_ERROR_UNEXPECTED);
+                }
+            }
+        }
+        None => {
+            // TODO: Is this error handling correct? Since,
+            // TODO: we already sent the request to cancel to Openfinex but
```

Comment:
What to do in this situation when the enclave cannot find the order in the orderbook storage.

Reply:
log a warning as enclave. return an error on jsonrpc to the caller of cancel function. continue without further action.

Reply:
Cool